homeassistant/core.py

```diff
@@ -2243,20 +2243,43 @@ def async_set(
 
         This method must be run in the event loop.
         """
-        new_state = str(new_state)
-        attributes = attributes or {}
-        old_state = self._states_data.get(entity_id)
-        if old_state is None:
-            # If the state is missing, try to convert the entity_id to lowercase
-            # and try again.
-            entity_id = entity_id.lower()
-            old_state = self._states_data.get(entity_id)
+        self.async_set_internal(
+            entity_id.lower(),
+            str(new_state),
+            attributes or {},
+            force_update,
+            context,
+            state_info,
+            timestamp or time.time(),
+        )
 
-        if old_state is None:
+    @callback
+    def async_set_internal(
+        self,
+        entity_id: str,
+        new_state: str,
+        attributes: Mapping[str, Any] | None,
+        force_update: bool,
+        context: Context | None,
+        state_info: StateInfo | None,
+        timestamp: float,
+    ) -> None:
+        """Set the state of an entity, add entity if it does not exist.
+
+        This method is intended to only be used by core internally
+        and should not be considered a stable API. We will make
+        breaking changes to this function in the future and it
+        should not be used in integrations.
+
+        This method must be run in the event loop.
+        """
+        if entity_id not in self._states_data:
+            old_state = None
             same_state = False
             same_attr = False
             last_changed = None
         else:
+            old_state = self._states_data[entity_id]
```

Comment:
Double lookups. Use a get with sentinel?

Reply:
calling `.get()` is actually slower

Reply:
Is that generically so? The positive path here is hot path. So we will always do that. Seems weird to be slower. 

We could catch the keyerror instead?

Reply:
Actually the `get()` will be faster because its going to be a hit most of the time.  I was thinking about the miss case

Reply:
miss case is faster
```
import timeit

x = {}


print(timeit.timeit('x.get("dog")', globals=globals()))
print(timeit.timeit('"dog" in x and x["dog"]', globals=globals()))
```

0.011831625015474856
0.01080325001385063

Reply:
Think keyerror would be even faster if we are optimizing :). Miss path is once in a lifetime

Reply:
yeah the try is faster

```
import timeit

x = {"dog":"cat"}

print(timeit.timeit('''
try:
    x["dog"]
except KeyError:
    pass
''', globals=globals()))
print(timeit.timeit('x.get("dog")', globals=globals()))
print(timeit.timeit('"dog" in x and x["dog"]', globals=globals()))
```

```
0.011381832999177277
0.013859875005437061
0.020063042000401765
```